File: 0143-reorder-list/0143-reorder-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

/**
 Do not return anything, modify head in-place instead.
 */
function reorderList(head: ListNode | null): void {
    let fast = head 
    let slow = head 
    while (fast && fast.next) {
        slow = slow.next
        fast = fast.next.next
    }

    let curr = slow.next
    slow.next = null

    let prev = null
    while (curr) {
        let next = curr.next
        curr.next = prev
        prev = curr 
        curr = next
    }
    curr = head 
    while (curr && prev) {
        let next1 = curr.next
        let next2 = prev.next
        curr.next = prev
        prev.next = next1
        curr = next1
        prev = next2
    }
};

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        def SplitList(head):
            fast = head
            slow = head

            while fast.next:
                fast = fast.next

                if fast.next:
                    fast = fast.next
                    slow = slow.next

            sec = slow.next
            slow.next = None

            return sec

        def ReverseList(head):
            prev = None
            curr = head

            while curr:
                nxt = curr.next
                curr.next = prev
                prev = curr
                curr = nxt
                
            return prev
            
        mid = SplitList(head)
        reverse = ReverseList(mid)
        curr = head

        while reverse:
            nxt = curr.next
            curr.next = reverse
            reverse = reverse.next
            curr.next.next = nxt
            curr = curr.next.next